boot/freeldr/freeldr/arch/i386/multiboot.S

```diff
@@ -140,31 +139,29 @@ mb1:
 
 mb2:
     /* See if the boot device was passed in */
-    test dword ptr ds:[ebx + MB_INFO_FLAGS_OFFSET], MB_INFO_FLAG_BOOT_DEVICE
+    test dword ptr [rbx + MB_INFO_FLAGS_OFFSET], MB_INFO_FLAG_BOOT_DEVICE
 
     /* If no boot device known, assume first partition of first harddisk */
     mov dx, HEX(0180)
     jz mb3
 
     /* Load boot drive into DL, boot partition into DH */
-    mov edx, dword ptr ds:[ebx + MB_INFO_BOOT_DEVICE_OFFSET]
+    mov rdx, [rbx + MB_INFO_BOOT_DEVICE_OFFSET]
     bswap edx
-    inc dh
 
 mb3:
     /* Relocate itself to lower address */
     mov esi, INITIAL_BASE
     mov edi, FREELDR_BASE
-    mov ecx, offset __bss_start__ - FREELDR_BASE
-    shr ecx, 2
-    rep movsd
+    mov ecx, offset __bss_end__ - FREELDR_BASE
+    rep movsb
 
     /* Load segment registers for real-address mode */
-#ifdef _USE_ML
-    lgdt fword ptr ds:[gdtptr]
-#else
-    lgdt ds:[gdtptr]
-#endif
+    mov edi, offset gdtptr
+    lgdt lXdtPrefix [rdi]
```

Comment:
It doesn't work with `lgdt [gdtptr]`?

Reply:
No. it's doens't work.

Reply:
Weird. Because when compiling x64 code, the `lgdt [gdtptr]` works usually.... except with GAS, where one needs to do a trick:
https://git.reactos.org/?p=reactos.git;a=blob;f=boot/freeldr/freeldr/arch/uefi/amd64/uefiasm.S;hb=30952992e1a13b8f0c0bb4cabfaffe4ef3451522#l47
(this seems to be necessary only on x64; on x86 this is not necessary as it seems:
https://git.reactos.org/?p=reactos.git;a=blob;f=boot/freeldr/freeldr/arch/uefi/i386/uefiasm.S;hb=30952992e1a13b8f0c0bb4cabfaffe4ef3451522#l48 )

Reply:
I'll have a look at this / the other instances later to see, if I can improve it. For now it's fine.